### ISSUE TITLE
Cypress issues when using window.msw

### ISSUE THREAD

**Issue by robcaldecott** [kettanaito/msw#1052]
## Environment

```
msw@0.36.3
cypress@9.2.0
```

## Current behavior

I am trying to add mocks to a Cypress test based on the example here: https://mswjs.io/docs/api/setup-worker/use#examples
However, Cypress errors with `Cannot destructure property 'worker' of 'window.msw' as it is undefined.`

Sample repo here: https://github.com/robcaldecott/cypress-msw

```
npm install
npm run dev
```

In a second command shell:

```
npx cypress run
```

MSW is working as the app runs with a correctly mocked response. However, in my Cypress test I am trying to override the response based on the MSW docs, e.g.:

```ts
import { DefaultRequestBody, PathParams } from "msw";

describe("app", () => {
  beforeEach(() => {
    cy.visit("/");
  });

  it("renders names", () => {
    cy.window().then((window) => {
      // This will fail
      const { worker, rest } = window.msw;
      worker.use(
        rest.get<DefaultRequestBody, PathParams, string[]>(
          "/api/names",
          (req, res, ctx) => {
            return res(ctx.json(["Bob", "Alice"]));
          }
        )
      );
    });
    cy.findByLabelText(/loading/i).should("exist");
    cy.findByText(/bob/i).should("exist");
    cy.findByText(/alice/i).should("exist");
  });
});
```

However, `window.msw` exists when you open the devtools and inspect both the app running in a standalone browser window and the app running inside the Cypress UI.

I'm not sure what is going on here: either Cypress is passing the wrong `window` or there is some timing issue between the app starting, the Cypress test running and MSW being initialised.

So while this may turn out to be a bug with Cypress I don't know where else to start.

Am I missing something obvious?

## Screenshots

![image](https://user-images.githubusercontent.com/796702/148813600-f9a9824a-be7c-4529-a65e-4a4b63142ffd.png)

**Comment by robcaldecott** [comment]
OK, to make this work you need to signal when your app is ready (after starting the MSW server) and wait for this in your tests, e.g.:

main.tsx

```js
import("./mocks/browser").then(({ worker }) => {
  worker.start({ onUnhandledRequest: "bypass" });

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        refetchIntervalInBackground: false,
        retry: false,
      },
    },
  });

  // The secret sauce
  if (window.Cypress) {
    window.appReady = true;
  }

  ReactDOM.render(
    <React.StrictMode>
      <QueryClientProvider client={queryClient}>
        <App />
      </QueryClientProvider>
    </React.StrictMode>,
    document.getElementById("root")
  );
});
```

And in your tests:

```ts
beforeEach(() => {
  cy.visit("/");
  // Wait for MSW server to start
  cy.window().should("have.property", "appReady", true);
});
```

I will open a PR to update the MSW docs if possible.

**Comment by kettanaito** [comment]
Hey, @robcaldecott. Thank you for raising this. 

Yes, you're right, service worker registration takes time and creates a race condition between it and your app mounting/tests running. Awaiting it is the way to go. 

The way _how_ we await it can be discussed, though. You can introduce an intermediate state to your app (as you did above) but I think it's somewhat intrusive and makes you write test-oriented code as a part of your app (which is something we try to avoid). 

Did you consider assigning the worker registration promise to `window` instead?

```js
window.isMockReady = worker.start({ ... })
```

Since the worker registration returns a Promise, you will be able to await it in your Cypress setup before running tests:

```js
await win.isMockReady
```

I suppose, ideally, it'd be great to utilize something like a Cypress plugin, which I think I saw one of the community members doing. Here's their setup:

```js

// cypress/supports/index.js
import { worker } from '../../src/mocks/browser.js'

Cypress.on('test:before:run:async', async () => {
  // Start the worker before all tests in Cypress.
  await worker.start()
})
```

In their case, they are starting the worker as a part of the Cypress run. This is not mandatory, though, I'd recommend starting the worker within your app depending on the environmental flag (as you do). In the `test:before:run:async` hook you can await the `window.isMockReady`.

**Comment by robcaldecott** [comment]
@kettanaito Thanks for the suggestions. I don't think you can access `window` in a `Cypress.on` hook so I think I need to stick with `beforeEach` (but I am no Cypress expert).

I thought I'd try this instead in `main.tsx`:

```
worker.start({ onUnhandledRequest: "bypass" }).then(() => {
    if (window.Cypress) {
      window.appReady = true;
    }
  });
```

But this cause Cypress to fail:

```
expected <window> to have property appReady of true
```

Which is odd as if I `console.log` the `worker.start` code it's definitely adding that flag.

**Comment by robcaldecott** [comment]
This looks slightly more elegant as it waits for the worker to start before doing anything.

```tsx
import("./mocks/browser").then(({ worker }) => {
  worker.start({ onUnhandledRequest: "bypass" }).then(() => {
    if (window.Cypress) {
      window.appReady = true;
    }

    const queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          // Disable automatic retries
          retry: false,
        },
      },
    });

    ReactDOM.render(
      <React.StrictMode>
        <QueryClientProvider client={queryClient}>
          <App />
        </QueryClientProvider>
      </React.StrictMode>,
      document.getElementById("root")
    );
  });
});
```

But it doesn't always work and I get random Cypress assertions when using `npx cypress run` which looks like the custom MSW handlers are installed but not used.

Better MSW integration with Cypress would definitely be good.

**Comment by robcaldecott** [comment]
I also have issues running Cypress in a docker container pointing at an app running locally: the MSW overrides in the tests just don't seem to work.

E.g. start the app:

```
npm run dev
```

Then in another console run docker:

```
docker run -it -v $PWD:/e2e -w /e2e cypress/included:9.2.1 --config baseUrl=http://host.docker.internal:3000,video=false
```

But tests that use `worker.use` fail.

No idea how to debug this but I suspect a timing issue of some kind.

**Comment by nemonemi** [comment]
I have the same problem.
For me, I have access to the `window` in the `Cypress.on('test:before:run:async'`, however, it doesn't contain the `msw` or `isMockReady` because it got executed before the application was initialized and the `window` extended.

```ts
// browser.ts
import { setupWorker } from 'msw';
import { handlers } from './handlers';

// This configures a Service Worker with the given request handlers.
export const worker = setupWorker(...handlers);

window.isMockReady = worker.start();

window.msw = {
  worker,
};
```

```ts
// e2e/src/support/index.ts

Cypress.on('test:before:run:async', async () => {
  console.log(window);
  
  await window.isMockReady;
});
```

```ts
// test file
  beforeEach(() => {
    cy.window().then((window) => {
      const { worker } = window.msw;

      worker.resetHandlers();
    });
```
